Typechecking fails with "moduleResolution": "node16" or "nodenext"

### Initial checklist

- [X] I read the [support docs](https://github.com/syntax-tree/.github/blob/main/support.md)
- [X] I read the [contributing guide](https://github.com/syntax-tree/.github/blob/main/contributing.md)
- [X] I agree to follow the [code of conduct](https://github.com/syntax-tree/.github/blob/main/code-of-conduct.md)
- [X] I searched [issues](https://github.com/search?q=user%3Asyntax-tree&type=Issues) and couldn’t find anything (or linked relevant results below)

### Affected packages and versions

mdast-util-directive@2.2.2

### Link to runnable example

https://github.com/jameshfisher/bug-report-remark-node16

### Steps to reproduce

Set `"moduleResolution"` to `"node16"` or `"nodenext"` in `tsconfig.json`.

Run `tsc`.

### Expected behavior

No typechecking errors.

### Actual behavior

```
$ npm run build

> test-remark-node16@1.0.0 build
> tsc

example.ts:10:10 - error TS2367: This comparison appears to be unintentional because the types '"link" | "html" | "strong" | "image" | "text" | "linkReference" | "emphasis" | "delete" | "break" | "footnote" | "inlineCode" | "imageReference" | "footnoteReference"' and '"textDirective"' have no overlap.

10   return p.type === "textDirective";
            ~~~~~~~~~~~~~~~~~~~~~~~~~~

node_modules/mdast-util-directive/index.d.ts:14:69 - error TS2694: Namespace '"/Users/jim/dev/tmp/test-remark-node16/node_modules/mdast-util-to-markdown/lib/types"' has no exported member 'Context'.

14 export type Context = import('mdast-util-to-markdown/lib/types.js').Context
                                                                       ~~~~~~~

node_modules/mdast-util-directive/index.d.ts:17:41 - error TS2834: Relative import paths need explicit file extensions in EcmaScript imports when '--moduleResolution' is 'node16' or 'nodenext'. Consider adding an extension to the import path.

17 export type ContainerDirective = import('./complex-types').ContainerDirective
                                           ~~~~~~~~~~~~~~~~~

node_modules/mdast-util-directive/index.d.ts:18:36 - error TS2834: Relative import paths need explicit file extensions in EcmaScript imports when '--moduleResolution' is 'node16' or 'nodenext'. Consider adding an extension to the import path.

18 export type LeafDirective = import('./complex-types').LeafDirective
                                      ~~~~~~~~~~~~~~~~~

node_modules/mdast-util-directive/index.d.ts:19:36 - error TS2834: Relative import paths need explicit file extensions in EcmaScript imports when '--moduleResolution' is 'node16' or 'nodenext'. Consider adding an extension to the import path.

19 export type TextDirective = import('./complex-types').TextDirective
                                      ~~~~~~~~~~~~~~~~~


Found 5 errors in 2 files.

Errors  Files
     1  example.ts:10
     4  node_modules/mdast-util-directive/index.d.ts:14
```

### Affected runtime and version

Node 16

### Affected package manager and version

npm 8

### Affected OS and version

macOS

### Build and bundle tools

_No response_

--- a/index.js
+++ b/index.js
@@ -7,7 +7,7 @@
  * @typedef {import('mdast-util-from-markdown').CompileContext} CompileContext
  * @typedef {import('mdast-util-from-markdown').Token} Token
  * @typedef {import('mdast-util-to-markdown/lib/types.js').Handle} ToMarkdownHandle
- * @typedef {import('mdast-util-to-markdown/index.js').Context} Context
+ * @typedef {import('mdast-util-to-markdown').Context} Context
  * @typedef {import('mdast-util-to-markdown/lib/types.js').Options} ToMarkdownExtension
  * @typedef {import('./complex-types.js').ContainerDirective} ContainerDirective
  * @typedef {import('./complex-types.js').LeafDirective} LeafDirective

--- a/index.js
+++ b/index.js
@@ -398,7 +398,7 @@ function fence(node) {
 
   return ':'.repeat(size)
 
-  /** @type {import('unist-util-visit-parents/complex-types.js').BuildVisitor<Root, Directive>} */
+  /** @type {import('unist-util-visit-parents').BuildVisitor<Root, Directive>} */
   function onvisit(_, parents) {
     let index = parents.length
     let nesting = 0

--- a/index.js
+++ b/index.js
@@ -9,9 +9,9 @@
  * @typedef {import('mdast-util-to-markdown/lib/types.js').Handle} ToMarkdownHandle
  * @typedef {import('mdast-util-to-markdown/lib/types.js').Context} Context
  * @typedef {import('mdast-util-to-markdown/lib/types.js').Options} ToMarkdownExtension
- * @typedef {import('./complex-types').ContainerDirective} ContainerDirective
- * @typedef {import('./complex-types').LeafDirective} LeafDirective
- * @typedef {import('./complex-types').TextDirective} TextDirective
+ * @typedef {import('./complex-types.js').ContainerDirective} ContainerDirective
+ * @typedef {import('./complex-types.js').LeafDirective} LeafDirective
+ * @typedef {import('./complex-types.js').TextDirective} TextDirective
  * @typedef {ContainerDirective|LeafDirective|TextDirective} Directive
  */
 

--- a/index.js
+++ b/index.js
@@ -7,7 +7,7 @@
  * @typedef {import('mdast-util-from-markdown').CompileContext} CompileContext
  * @typedef {import('mdast-util-from-markdown').Token} Token
  * @typedef {import('mdast-util-to-markdown/lib/types.js').Handle} ToMarkdownHandle
- * @typedef {import('mdast-util-to-markdown/lib/types.js').Context} Context
+ * @typedef {import('mdast-util-to-markdown/index.js').Context} Context
  * @typedef {import('mdast-util-to-markdown/lib/types.js').Options} ToMarkdownExtension
  * @typedef {import('./complex-types.js').ContainerDirective} ContainerDirective
  * @typedef {import('./complex-types.js').LeafDirective} LeafDirective
@@ -398,7 +398,7 @@ function fence(node) {
 
   return ':'.repeat(size)
 
-  /** @type {import('unist-util-visit-parents/complex-types').BuildVisitor<Root, Directive>} */
+  /** @type {import('unist-util-visit-parents/complex-types.js').BuildVisitor<Root, Directive>} */
   function onvisit(_, parents) {
     let index = parents.length
     let nesting = 0

--- a/package.json
+++ b/package.json
@@ -38,10 +38,10 @@
   "dependencies": {
     "@types/mdast": "^3.0.0",
     "@types/unist": "^2.0.0",
-    "mdast-util-to-markdown": "^1.3.0",
+    "mdast-util-to-markdown": "^1.5.0",
     "parse-entities": "^4.0.0",
     "stringify-entities": "^4.0.0",
-    "unist-util-visit-parents": "^5.0.0"
+    "unist-util-visit-parents": "^5.1.3"
   },
   "devDependencies": {
     "@types/tape": "^4.0.0",

--- a/package.json
+++ b/package.json
@@ -51,7 +51,6 @@
     "prettier": "^2.0.0",
     "remark-cli": "^11.0.0",
     "remark-preset-wooorm": "^9.0.0",
-    "rimraf": "^3.0.0",
     "tape": "^5.0.0",
     "type-coverage": "^2.0.0",
     "typescript": "^4.0.0",
@@ -59,10 +58,10 @@
     "xo": "^0.52.0"
   },
   "scripts": {
-    "build": "rimraf \"{index,test}.d.ts\" && tsc && type-coverage",
+    "build": "tsc --build --clean && tsc --build && type-coverage",
     "format": "remark . -qfo && prettier . -w --loglevel warn && xo --fix",
     "test-api": "node --conditions development test.js",
-    "test-coverage": "c8 --check-coverage --branches 100 --functions 100 --lines 100 --statements 100 --reporter lcov node --conditions development test.js",
+    "test-coverage": "c8 --check-coverage --100 --reporter lcov npm run test-api",
     "test": "npm run build && npm run format && npm run test-coverage"
   },
   "prettier": {

--- a/tsconfig.json
+++ b/tsconfig.json
@@ -4,7 +4,7 @@
     "target": "ES2020",
     "lib": ["ES2020"],
     "module": "ES2020",
-    "moduleResolution": "node",
+    "moduleResolution": "node16",
     "allowJs": true,
     "checkJs": true,
     "declaration": true,

--- a/tsconfig.json
+++ b/tsconfig.json
@@ -1,16 +1,16 @@
 {
   "include": ["*.js"],
   "compilerOptions": {
-    "target": "ES2020",
-    "lib": ["ES2020"],
-    "module": "ES2020",
-    "moduleResolution": "node16",
-    "allowJs": true,
     "checkJs": true,
     "declaration": true,
     "emitDeclarationOnly": true,
-    "allowSyntheticDefaultImports": true,
+    "exactOptionalPropertyTypes": true,
+    "forceConsistentCasingInFileNames": true,
+    "lib": ["es2020"],
+    "module": "node16",
+    "newLine": "lf",
     "skipLibCheck": true,
-    "strict": true
+    "strict": true,
+    "target": "es2020"
   }
 }

